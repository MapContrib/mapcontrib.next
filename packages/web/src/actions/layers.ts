import { ILayer, ILayerSource } from 'types';

export enum Actions {
  ADD_LAYER = 'ADD_LAYER',
  ADD_SOURCE_TO_LAYER = 'ADD_SOURCE_TO_LAYER',
  REMOVE_SOURCE_FROM_LAYER = 'REMOVE_SOURCE_FROM_LAYER',
  EDIT_LAYER = 'EDIT_LAYER',
  REMOVE_LAYER = 'REMOVE_LAYER',
}

export type Action =
  | {
      layer: ILayer;
      type: Actions.ADD_LAYER;
    }
  | {
      id: string;
      source: ILayerSource;
      type: Actions.ADD_SOURCE_TO_LAYER;
    }
  | {
      id: string;
      source: ILayerSource;
      type: Actions.REMOVE_SOURCE_FROM_LAYER;
    }
  | {
      data: Partial<ILayer>;
      id: string;
      type: Actions.EDIT_LAYER;
    }
  | {
      id: string;
      type: Actions.REMOVE_LAYER;
    };

export const addLayer = (layer: ILayer): Action => ({
  layer,
  type: Actions.ADD_LAYER,
});

export const addSourceToLayer = (layer: ILayer, source: ILayerSource): Action =>
  addSourceToLayerById(layer.id, source);

export const addSourceToLayerById = (
  id: string,
  source: ILayerSource
): Action => ({
  id,
  source,
  type: Actions.ADD_SOURCE_TO_LAYER,
});

export const removeSourceFromLayer = (
  layer: ILayer,
  source: ILayerSource
): Action => removeSourceFromLayerById(layer.id, source);

export const removeSourceFromLayerById = (
  id: string,
  source: ILayerSource
): Action => ({
  id,
  source,
  type: Actions.REMOVE_SOURCE_FROM_LAYER,
});

export const editLayer = (layer: ILayer, data: ILayer): Action =>
  editLayerById(layer.id, data);

export const editLayerById = (id: string, data: Partial<ILayer>): Action => ({
  data,
  id,
  type: Actions.EDIT_LAYER,
});

export const removeLayer = (layer: ILayer): Action => removeLayerById(layer.id);

export const removeLayerById = (id: string): Action => ({
  id,
  type: Actions.REMOVE_LAYER,
});
