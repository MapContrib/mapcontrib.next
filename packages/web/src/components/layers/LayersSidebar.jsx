import React from 'react';
import PropTypes from 'prop-types';
import shortid from 'shortid';
import styled from 'styled-components';
import { RedTheme, DefaultTheme, Sidebar, List } from 'osm-ui-react';

import LayerCreationForm from 'components/layers/LayerCreationForm';

const StyledDiv = styled.div`
  .layers {
    padding: 0;
    margin: 0;
    margin-bottom: 5rem;

    li {
      display: flex;
      justify-content: space-between;
    }
  }
`;

class LayerSidebar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isCreating: false
    };

    this.toggleCreationMode = this.toggleCreationMode.bind(this);
    this.createLayer = this.createLayer.bind(this);
  }

  toggleCreationMode() {
    this.setState({
      isCreating: !this.state.isCreating
    });
  }

  goToLayer(id) {
    const { history, match } = this.props;

    const newPath = `${match && match.url}/${id}`;
    history.push(newPath);
  }

  createLayer(layer) {
    const id = shortid.generate();

    this.props.addLayer({
      ...layer,
      id
    });

    this.toggleCreationMode();
    this.goToLayer(id);
  }

  render() {
    const { history, match, themePath, layers, ...props } = this.props;

    const layerItems = layers.map(layer => (
      <List.Item key={layer.id} onClick={() => this.goToLayer(layer.id)}>
        <div>{layer.name}</div>
        <div>{layer.type}</div>
      </List.Item>
    ));

    return (
      <StyledDiv>
        <RedTheme>
          <Sidebar
            opened={!!match}
            position="right"
            title="Layers"
            onClickClose={() => history.replace(themePath)}
            {...props}
          >
            <DefaultTheme>
              <List className="layers">{layerItems}</List>
            </DefaultTheme>
            <Sidebar.Nav>
              <ul>
                {/* <li>
                  <Link to="">Add a missing point</Link>
                </li> */}
                <li>
                  <a onClick={this.toggleCreationMode}>Add a data layer</a>
                  {this.state.isCreating && (
                    <LayerCreationForm createLayer={this.createLayer} />
                  )}
                </li>
              </ul>
            </Sidebar.Nav>
          </Sidebar>
        </RedTheme>
      </StyledDiv>
    );
  }
}

LayerSidebar.propTypes = {
  layers: PropTypes.array,
  addLayer: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  match: PropTypes.object,
  themePath: PropTypes.string.isRequired
};

LayerSidebar.defaultProps = {
  layers: []
};

LayerSidebar.displayName = 'LayerSidebar';

export default LayerSidebar;
