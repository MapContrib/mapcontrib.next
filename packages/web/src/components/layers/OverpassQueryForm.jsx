import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Form, Button } from 'osm-ui-react';

const StyledDiv = styled.div`
  button {
    display: block;
    width: 50%;
    margin: 1rem 0;
    margin-left: 50%;
  }
`;

class OverpassQueryForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      query: props.query
    };

    this.submit = this.submit.bind(this);
  }

  submit() {
    this.props.updateQuery(this.state.query);
  }

  render() {
    return (
      <Fragment>
        <StyledDiv>
          <Form.Label htmlFor="query">Overpass query</Form.Label>
          <Form.Textarea
            id="query"
            placeholder="Enter overpass query..."
            autoFocus={false}
            labelKey="query"
            valueKey="id"
            onChange={e => this.setState({ query: e.target.value })}
            value={this.state.query}
          />
          <Button type="submit" onClick={this.submit}>
            Update query
          </Button>
        </StyledDiv>
      </Fragment>
    );
  }
}

OverpassQueryForm.propTypes = {
  query: PropTypes.string,
  updateQuery: PropTypes.func.isRequired
};

OverpassQueryForm.defaultProps = {
  query: ''
};

OverpassQueryForm.displayName = 'OverpassQueryForm';

export default OverpassQueryForm;
