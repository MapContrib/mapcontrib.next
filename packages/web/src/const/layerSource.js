export const OSMOSE_SOURCE = 'OSMOSE_SOURCE';
export const OVERPASS_SOURCE = 'OVERPASS_SOURCE';
