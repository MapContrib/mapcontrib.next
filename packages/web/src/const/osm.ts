export const CHANGESET_CREATED_BY = 'MapContrib {version}';
export const CHANGESET_COMMENT =
  'Contribution sent from the following MapContrib theme: {url}';
