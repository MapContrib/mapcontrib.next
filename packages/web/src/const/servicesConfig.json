{
  "OSMOSE_PROD_URL": "https://opendata.osmose.openstreetmap.fr/fr/api/0.2",
  "OVERPASS_PROD_URL": "https://overpass-api.de/api",

  "OSM_PROD_URL": "https://www.openstreetmap.org",
  "OSM_DEV_URL": "https://api06.dev.openstreetmap.org",

  "OAUTH_SECRET": "aaQcLOJFno7EGM0hC8nXaE85zOY4UP9uxklqpo7O",
  "OAUTH_CONSUMER_KEY": "WiLOF1hp5Aloyldyd4VLN2ZjBvqoyUNO3n2mDCpL"
}
