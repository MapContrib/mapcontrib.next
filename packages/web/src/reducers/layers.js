import {
  ADD_LAYER,
  ADD_SOURCE_TO_LAYER,
  REMOVE_SOURCE_FROM_LAYER,
  EDIT_LAYER,
  REMOVE_LAYER
} from 'actions/layers';

export const initialState = {};

const blackList = ['id', 'type', 'sources'];

export default function layers(state = initialState, action = { type: null }) {
  switch (action.type) {
    case ADD_LAYER:
      action.layer.isVisible = true;
      action.layer.sources = [];

      const newState = {
        ...state,
        [action.layer.id]: action.layer
      };

      return newState;

    case ADD_SOURCE_TO_LAYER:
      if (!state[action.id]) {
        throw new LayerException(`Layer not found (ID: ${action.id})`);
      }

      const layer = { ...state[action.id] };
      const source = action.source;

      if (layer.type && layer.type !== source.type) {
        throw new LayerException(
          `Source type ${source.type} does not match layer type: ${layer.type})`
        );
      }

      if (!layer.type) layer.type = source.type;

      let newSources;

      if (layer.sources) {
        newSources = [...layer.sources];
        newSources.push(source.id);
      } else newSources = [source.id];

      layer.sources = newSources;

      return {
        ...state,
        [action.id]: layer
      };

    case REMOVE_SOURCE_FROM_LAYER:
      if (!state[action.id]) {
        throw new LayerException(`Layer not found (ID: ${action.id})`);
      }

      const layerToRemoveFrom = { ...state[action.id] };
      const sourceToRemove = action.source;

      const sources = new Set(layerToRemoveFrom.sources);
      sources.delete(sourceToRemove.id);

      layerToRemoveFrom.sources = Array.from(sources);

      return {
        ...state,
        [action.id]: layerToRemoveFrom
      };

    case EDIT_LAYER:
      const keys = Object.keys(action.data);

      blackList.forEach(item => {
        if (keys.includes(item))
          throw new LayerException(
            `Attempt to modifify blacklisted layer key (key: ${item})`
          );
      });

      const editedLayer = {
        ...state[action.id],
        ...action.data
      };

      return {
        ...state,
        [action.id]: editedLayer
      };

    case REMOVE_LAYER:
      delete state[action.id];
      return { ...state };

    default:
      return state;
  }
}

export function LayerException(message) {
  this.message = message;
  this.name = 'LayerException';
}
