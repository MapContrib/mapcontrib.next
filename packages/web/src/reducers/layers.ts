import { Action, Actions } from 'actions/layers';
import LayerException from 'exceptions/layer';
import { ILayer } from 'types';

export interface IState {
  [key: string]: ILayer;
}

const blackList = ['id', 'type', 'sources'];

export default function layers(state: IState = {}, action: Action) {
  switch (action.type) {
    case Actions.ADD_LAYER:
      action.layer.isVisible = true;
      action.layer.sources = [];

      const newState = {
        ...state,
        [action.layer.id]: action.layer,
      };

      return newState;

    case Actions.ADD_SOURCE_TO_LAYER:
      if (!state[action.id]) {
        throw new LayerException(`Layer not found (ID: ${action.id})`);
      }

      const layer = { ...state[action.id] };
      const source = action.source;

      if (layer.type && layer.type !== source.type) {
        throw new LayerException(
          `Source type ${source.type} does not match layer type: ${layer.type})`
        );
      }

      if (!layer.type) {
        layer.type = source.type;
      }

      let newSources;

      if (layer.sources) {
        newSources = [...layer.sources];
        newSources.push(source.id);
      } else {
        newSources = [source.id];
      }

      layer.sources = newSources;

      return {
        ...state,
        [action.id]: layer,
      };

    case Actions.REMOVE_SOURCE_FROM_LAYER:
      if (!state[action.id]) {
        throw new LayerException(`Layer not found (ID: ${action.id})`);
      }

      const layerToRemoveFrom = { ...state[action.id] };
      const sourceToRemove = action.source;

      const sources = new Set(layerToRemoveFrom.sources);
      sources.delete(sourceToRemove.id);

      layerToRemoveFrom.sources = Array.from(sources);

      return {
        ...state,
        [action.id]: layerToRemoveFrom,
      };

    case Actions.EDIT_LAYER:
      const keys = Object.keys(action.data);

      blackList.forEach(item => {
        if (keys.includes(item)) {
          throw new LayerException(
            `Attempt to modifify blacklisted layer key (key: ${item})`
          );
        }
      });

      const editedLayer = {
        ...state[action.id],
        ...action.data,
      };

      return {
        ...state,
        [action.id]: editedLayer,
      };

    case Actions.REMOVE_LAYER:
      delete state[action.id];
      return { ...state };

    default:
      return state;
  }
}
