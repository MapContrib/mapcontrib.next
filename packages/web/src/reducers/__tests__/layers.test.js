import {
  initialState,
  default as reducer,
  LayerException
} from 'reducers/layers';
import {
  addLayer,
  addSourceToLayerById,
  removeLayerById,
  removeSourceFromLayerById,
  editLayerById
} from 'actions/layers';
import { sourceTypes } from 'const/layers';

const defaultLayerAttributes = {
  id: 'id',
  name: 'name',
  description: 'description',
  isVisible: true,
  representationType: 'marker',
  sources: []
};

const overpassLayer = {
  ...defaultLayerAttributes,
  id: 'UUID_OVERPASS'
};

const osmoseLayer = {
  ...defaultLayerAttributes,
  id: 'UUID_OSMOSE'
};

const overpassSource = {
  id: 'UUID_OVERPASS_SOURCE',
  type: sourceTypes.OVERPASS,
  origin: 'origin',
  features: []
};

const osmoseSource = {
  id: 'UUID_OSMOSE_SOURCE',
  type: sourceTypes.OSMOSE,
  origin: 'origin',
  features: []
};

describe('Layers reducer', () => {
  it('Should return the initial state', () => {
    expect(reducer()).toBe(initialState);
  });

  it('Should be able to add a layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));

    expect(state).toMatchSnapshot();

    const state2 = reducer(state, addLayer(osmoseLayer));

    expect(state2).toMatchSnapshot();
    expect(Object.keys(state2).length).toBe(2);
    expect(state2).not.toBe(state);
  });

  it('Should be able to remove a layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const state2 = reducer(state, addLayer(osmoseLayer));

    const state3 = reducer(state2, removeLayerById(overpassLayer.id));

    expect(Object.keys(state3).length).toBe(1);
    expect(state3).toMatchSnapshot();
    expect(state3).not.toBe(state2);
  });

  it('Should be able to add sources to a layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const state2 = reducer(
      state,
      addSourceToLayerById(overpassLayer.id, overpassSource)
    );

    expect(state2).toMatchSnapshot();
    expect(state2[overpassLayer.id].type).toBe(overpassSource.type);
    expect(state2[overpassLayer.id].sources).toHaveLength(1);
    expect(state2[overpassLayer.id].sources[0]).toBe(overpassSource.id);
    expect(state2).not.toBe(state);

    const state3 = reducer(
      state2,
      addSourceToLayerById(overpassLayer.id, {
        id: 'newSource',
        type: sourceTypes.OVERPASS
      })
    );

    expect(state3).toMatchSnapshot();
    expect(state3[overpassLayer.id].sources).toHaveLength(2);
    expect(state3[overpassLayer.id].sources[1]).toBe('newSource');
    expect(state3).not.toBe(state);
  });

  it('Should be able to remove a source from a layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const state2 = reducer(
      state,
      addSourceToLayerById(overpassLayer.id, overpassSource)
    );
    const state3 = reducer(
      state2,
      removeSourceFromLayerById(overpassLayer.id, overpassSource)
    );

    expect(state3).toMatchSnapshot();
    expect(state3[overpassLayer.id].sources).toHaveLength(0);
    expect(state3).not.toBe(state);
  });

  it('Should be able to edit a layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));

    const newData = {
      name: 'newName',
      description: 'newDescription',
      isVisible: 'newIsVisible',
      representationType: 'newRepresentationType'
    };

    const state2 = reducer(state, editLayerById(overpassLayer.id, newData));

    const editedLayer = state2[overpassLayer.id];

    expect(state2).toMatchSnapshot();
    Object.keys(newData).forEach(key =>
      expect(editedLayer[key]).toBe(newData[key])
    );
    expect(state2).not.toBe(state);
  });

  it('Should throw an exception when adding a source to an unknown layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const addSource = () =>
      reducer(state, addSourceToLayerById('Unknown_ID', overpassSource));

    expect(addSource).toThrow(LayerException);
  });

  it('Should throw an exception when removing a source from an unknown layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const removeSource = () =>
      reducer(state, removeSourceFromLayerById('Unknown_ID', overpassSource));

    expect(removeSource).toThrow(LayerException);
  });

  it('Should throw an exception when adding a source to a layer of different type', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const state2 = reducer(
      state,
      addSourceToLayerById(overpassLayer.id, overpassSource)
    );

    const addSource = () => {
      reducer(state2, addSourceToLayerById(overpassLayer.id, osmoseSource));
    };

    expect(addSource).toThrow(LayerException);
  });

  it('Should throw an exception when editing a blacklisted item in a layer', () => {
    const state = reducer(initialState, addLayer(overpassLayer));
    const editLayer = () =>
      reducer(state, editLayerById(overpassLayer.id, { id: 2 }));

    expect(editLayer).toThrow(LayerException);
  });
});
