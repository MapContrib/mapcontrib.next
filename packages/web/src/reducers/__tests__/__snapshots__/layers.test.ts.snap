// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layers reducer Should be able to add a layer 1`] = `
Object {
  "UUID_OVERPASS": Object {
    "description": "description",
    "id": "UUID_OVERPASS",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [],
    "type": "OVERPASS",
  },
}
`;

exports[`Layers reducer Should be able to add a layer 2`] = `
Object {
  "UUID_OSMOSE": Object {
    "description": "description",
    "id": "UUID_OSMOSE",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [],
    "type": "OSMOSE",
  },
  "UUID_OVERPASS": Object {
    "description": "description",
    "id": "UUID_OVERPASS",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [],
    "type": "OVERPASS",
  },
}
`;

exports[`Layers reducer Should be able to add sources to a layer 1`] = `
Object {
  "UUID_OVERPASS": Object {
    "description": "description",
    "id": "UUID_OVERPASS",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [
      "UUID_OVERPASS_SOURCE",
    ],
    "type": "OVERPASS",
  },
}
`;

exports[`Layers reducer Should be able to add sources to a layer 2`] = `
Object {
  "UUID_OVERPASS": Object {
    "description": "description",
    "id": "UUID_OVERPASS",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [
      "UUID_OVERPASS_SOURCE",
      "newSource",
    ],
    "type": "OVERPASS",
  },
}
`;

exports[`Layers reducer Should be able to edit a layer 1`] = `
Object {
  "UUID_OVERPASS": Object {
    "description": "newDescription",
    "id": "UUID_OVERPASS",
    "isVisible": true,
    "name": "newName",
    "representationType": "HEATMAP",
    "sources": Array [],
    "type": "OVERPASS",
  },
}
`;

exports[`Layers reducer Should be able to remove a layer 1`] = `
Object {
  "UUID_OSMOSE": Object {
    "description": "description",
    "id": "UUID_OSMOSE",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [],
    "type": "OSMOSE",
  },
}
`;

exports[`Layers reducer Should be able to remove a source from a layer 1`] = `
Object {
  "UUID_OVERPASS": Object {
    "description": "description",
    "id": "UUID_OVERPASS",
    "isVisible": true,
    "name": "name",
    "representationType": "MARKER",
    "sources": Array [],
    "type": "OVERPASS",
  },
}
`;
